# Proceso EDA
---
1. Cargar los datos como un Pandas dataframe

2. LLevar a cabo un proceso EDA (Exploratory Data Analysis) para entender/comprender sus datos. Y esto significa:
  - Cuántas instancias y atributos hay
  - Qué tipo de atributos hay (numéricos o categóricos). Esto se hace para comprobar si hay características categóricas que deban ser codificadas (como dummies / one-hot-encoding)
  - Qué atributos tienen valores faltantes, y cuántos
  - Si se trata de un problema de clasificación o de regresión (variable de salida es una clase o un valor continuo), y en caso de clasificación, si la clase está desequilibrada.

3. Codificar el Pandas dataframe como una matriz numpy (transformar los valores categóricos y los valores faltantes)

4. Diseñar modelos de aprendizaje automático con scikit-learn

In [90]:
# Cargamos el archivo
import pandas as pd
data_df = pd.read_csv("attrition_25/attrition_availabledata_06.csv") 

Podemos comprobar el valos de las primeras instancias del dataframe

In [91]:
data_df.head()

hrs  absences  JobInvolvement  PerformanceRating  \
0  5.896431         5               3                  3   
1  6.114877        19               3                  3   
2  6.396730        11               3                  3   
3  5.722452        22               4                  3   
4  5.663001        21               3                  3   

   EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  Age  \
0                      2.0              4.0              3.0   33   
1                      3.0              3.0              3.0   20   
2                      1.0              4.0              4.0   46   
3                      1.0              1.0              4.0   40   
4                      4.0              4.0              3.0   38   

      BusinessTravel              Department  ...  Over18  PercentSalaryHike  \
0      Travel_Rarely  Research & Development  ...       Y                 13   
1      Travel_Rarely  Research & Development  ...       Y                 13   
2      Travel_Rarely  Research & Development  ...       Y                 11   
3      Travel_Rarely                   Sales  ...       Y                 13   
4  Travel_Frequently  Research & Development  ...       Y                 17   

  StandardHours  StockOptionLevel  TotalWorkingYears TrainingTimesLastYear  \
0             8                 0                6.0                     3   
1             8                 1                2.0                     2   
2             8                 1               28.0                     2   
3             8                 1                9.0                     2   
4             8                 3               10.0                     2   

   YearsAtCompany YearsSinceLastPromotion YearsWithCurrManager  Attrition  
0               5                       1                    4         No  
1               2                       2                    2         No  
2               7                       4                    3        Yes  
3               8                       7                    1         No  
4              10                       9                    9         No  

[5 rows x 31 columns]

El dataset es demasiado grande, por eso no podemos visualizarlo al completo

In [92]:
data_df

hrs  absences  JobInvolvement  PerformanceRating  \
0      5.896431         5               3                  3   
1      6.114877        19               3                  3   
2      6.396730        11               3                  3   
3      5.722452        22               4                  3   
4      5.663001        21               3                  3   
...         ...       ...             ...                ...   
2935   5.550183        18               3                  3   
2936   6.976802        21               3                  4   
2937  10.092581        12               3                  3   
2938   8.333744        18               2                  3   
2939   6.833922         9               1                  3   

      EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  Age  \
0                         2.0              4.0              3.0   33   
1                         3.0              3.0              3.0   20   
2                         1.0              4.0              4.0   46   
3                         1.0              1.0              4.0   40   
4                         4.0              4.0              3.0   38   
...                       ...              ...              ...  ...   
2935                      2.0              2.0              3.0   32   
2936                      3.0              4.0              2.0   51   
2937                      3.0              4.0              3.0   28   
2938                      2.0              4.0              3.0   24   
2939                      4.0              2.0              2.0   29   

         BusinessTravel              Department  ...  Over18  \
0         Travel_Rarely  Research & Development  ...       Y   
1         Travel_Rarely  Research & Development  ...       Y   
2         Travel_Rarely  Research & Development  ...       Y   
3         Travel_Rarely                   Sales  ...       Y   
4     Travel_Frequently  Research & Development  ...       Y   
...                 ...                     ...  ...     ...   
2935         Non-Travel  Research & Development  ...       Y   
2936      Travel_Rarely  Research & Development  ...       Y   
2937  Travel_Frequently                   Sales  ...       Y   
2938      Travel_Rarely                   Sales  ...       Y   
2939  Travel_Frequently  Research & Development  ...       Y   

      PercentSalaryHike StandardHours  StockOptionLevel  TotalWorkingYears  \
0                    13             8                 0                6.0   
1                    13             8                 1                2.0   
2                    11             8                 1               28.0   
3                    13             8                 1                9.0   
4                    17             8                 3               10.0   
...                 ...           ...               ...                ...   
2935                 11             8                 1               12.0   
2936                 22             8                 1               18.0   
2937                 11             8                 0                7.0   
2938                 15             8                 2                6.0   
2939                 18             8                 0                8.0   

     TrainingTimesLastYear  YearsAtCompany YearsSinceLastPromotion  \
0                        3               5                       1   
1                        2               2                       2   
2                        2               7                       4   
3                        2               8                       7   
4                        2              10                       9   
...                    ...             ...                     ...   
2935                     2               7                       0   
2936                     2               7                       0   
2937                     3               7

Ahora observamos cuantos y qué tipo de datos vamos a estar tratando


In [93]:
print('The shape of the data table is:')
print('===============================')
print(data_df.shape)
print()

print('The types of the attributes are:')
print('================================')
data_df.info()

print()

print('How many missing values per attribute:')
print('======================================')
print(data_df.isnull().sum())

print()

print('Fraction of missing values per attribute:')
print('======================================')
print(data_df.isnull().mean())


The shape of the data table is:
(2940, 31)

The types of the attributes are:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2940 entries, 0 to 2939
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   hrs                      2940 non-null   float64
 1   absences                 2940 non-null   int64  
 2   JobInvolvement           2940 non-null   int64  
 3   PerformanceRating        2940 non-null   int64  
 4   EnvironmentSatisfaction  2921 non-null   float64
 5   JobSatisfaction          2928 non-null   float64
 6   WorkLifeBalance          2913 non-null   float64
 7   Age                      2940 non-null   int64  
 8   BusinessTravel           2940 non-null   object 
 9   Department               2940 non-null   object 
 10  DistanceFromHome         2940 non-null   int64  
 11  Education                2940 non-null   int64  
 12  EducationField           2940 non-null   object 
 13  E

Gracias a este análisis sabemos que tendremos:

+ 2940 muestras (trabajadores)
+ 31 datos de cada una de ellas
    - Datos existentes de 3 tipos (object, int y float64)
    - Algunos datos faltan de ciertos trabajadores, pero comparándolo con el total no faltan demasiados
    - existen 8 variables de tipo object o categóricas que tendremos que convertir
+ Nuestra variable objetivo (attrition) es de tipo categórica, por lo que estamos ante un problema de clasificación


Antes de convertir las variables categóricas, debemos analizar cuales poseen una alta cardinalidad

In [94]:
# Paso 1: Identificar variables categóricas
categorical_columns = data_df.select_dtypes(include=['object']).columns

# Paso 2: Contar las categorías únicas por variable
cardinality = data_df[categorical_columns].nunique()

print("Cardinalidad de las variables categóricas:")
print(cardinality)

Cardinalidad de las variables categóricas:
BusinessTravel    3
Department        3
EducationField    6
Gender            2
JobRole           9
MaritalStatus     3
Over18            1
Attrition         2
dtype: int64


Éstos datos nos indican que no tenemos ninguna variable con demasiada cardinalidad.

variable con más cardinalidad: "JobRole" con 9

Podremos convertir nuestras variables categóricas a datos enteros.

In [96]:
from sklearn.preprocessing import LabelEncoder

# Crear una instancia de LabelEncoder
le = LabelEncoder()

#Identificamos las variables categóricas
categorical_columns = data_df.select_dtypes(include=['object']).columns

for column in categorical_columns:
    data_df[column] = le.fit_transform(data_df[column])  # Aplicar el LabelEncoder
    print(f"Mapping of categories to numbers for '{column}':")
    for i, category in enumerate(le.classes_):
        print(f"{category}: {i}")
    print() 

Mapping of categories to numbers for 'BusinessTravel':
Non-Travel: 0
Travel_Frequently: 1
Travel_Rarely: 2

Mapping of categories to numbers for 'Department':
Human Resources: 0
Research & Development: 1
Sales: 2

Mapping of categories to numbers for 'EducationField':
Human Resources: 0
Life Sciences: 1
Marketing: 2
Medical: 3
Other: 4
Technical Degree: 5

Mapping of categories to numbers for 'Gender':
Female: 0
Male: 1

Mapping of categories to numbers for 'JobRole':
Healthcare Representative: 0
Human Resources: 1
Laboratory Technician: 2
Manager: 3
Manufacturing Director: 4
Research Director: 5
Research Scientist: 6
Sales Executive: 7
Sales Representative: 8

Mapping of categories to numbers for 'MaritalStatus':
Divorced: 0
Married: 1
Single: 2

Mapping of categories to numbers for 'Over18':
Y: 0

Mapping of categories to numbers for 'Attrition':
No: 0
Yes: 1



Puesto que nos encontramos ante un problema de clasificación debemos analizar si nuestra variable Attrition está equilibrada:

In [ ]:

# Comprobar cuántas veces aparece cada clase en la columna 'Attrition'
print(data_df['Attrition'].value_counts())

# Comprobar la proporción de cada clase
print(data_df['Attrition'].value_counts(normalize=True))

Attrition
0    2466
1     474
Name: count, dtype: int64
Attrition
0    0.838776
1    0.161224
Name: proportion, dtype: float64


Observamos que está muy desequilibrada, por lo que